assignment2/BinaryTree.java

```diff
@@ -0,0 +1,198 @@
+
+        TreeNode(int item) {
+            value = item;
+        }
+
+        public int getValue() {
+            return value;
+        }
+
+        public TreeNode getLeft() {
+            return left;
+        }
+
+        @Override
+        public String toString() {
+            return String.format("TreeNode = %d", this.value);
+        }
+
+        public TreeNode getRight() {
+            return right;
+        }
+
+    }
+
+    /* Function to print all ancestors of purposed value
+    * @param node - root node
+    * @param value - value of the node on which we want to print ancestors
+    * @return true, if value exists in this tree, otherwise - false*/
+    boolean findAllAncestors(List<TreeNode> list, TreeNode node, int value)
+    {
+         //if tree is empty returns false
+        if (node == null) {
+            return false;
+        }
+
+        //if value exist in a tree
+        if (node.getValue() == value) {
+            return true;
+        }
+
+        //if value is present in left/right subtree
+        // then we'll print it
+        if (findAllAncestors(list, node.left, value) || findAllAncestors(list, node.right, value)) {
+            list.add(node);
+            return true;
+        }
+
+        //else return false
+        return false;
+    }
+
+    /*Traverse tree in preorder
+    * @param root - tree root*/
+    public void preorderTraversal (TreeNode root){ //метод для обхода в ширину
+        if (root != null){
+            System.out.print(root.getValue() + " ");
+            preorderTraversal (root.left);
+            preorderTraversal (root.right);
+        }
+    }
+
+    /*Populates tree with the array in order
+    * @param array - array which will fill the tree*/
+    public boolean populateTree(int[] array) {
+        for (int i = 0; i < array.length; i++) {
+            this.insert(array[i]);
+        }
+
+        return true;
+    }
+
+    /*@param value - value to insert*/
+    public void insert(int value) {
+        root = insert(root, value);
+    }
+
+    /*Inserts value to the tree
+    * @param node - root
+    * @param value - value to insert
+    * @return root*/
+    private static TreeNode insert(TreeNode node, int value) {
+        if (node == null) {
+            return new TreeNode(value);
+        }
+        if (value < node.value) {
+            node.left = insert(node.left, value);
+        } else if (value > node.value) {
+            node.right = insert(node.right, value);
+        }
+        return node;
+    }
+
+    /*Finds lowest common ancestor of two nodes
+    * @param root
+    * @param a - first node value
+    * @param b - second node value
+    * @return lowest common ancestor if both nodes are present
+    * in the tree, if one is not present - returns existing node, as LCA
+    * if both don't exist - returns null
+    * if a key is ancestor of other, then the ancestor key becomes LCA*/
+    public TreeNode findLowestCommonAncestor(TreeNode root, int a, int b) {

```

Comment:
Again, requiring user of the method to specify starting node makes it more difficult to use. Could you remove root parameter from public method? Then you can either have private helper method that does use root and is used for recursion (just like you did for insert - good job!), or you could move implementation to TreeNode.